### PR TITLE
Added source to ACR workflow

### LINKED ISSUE #646: [SPK] Pipeline Stage to build container image, push to ACR
**As a:** developer

**I want:** the auto generated azure-pipelines.yaml file from `spk project init` to automatically trigger ACR builds

**So that:** every commit to my project get a new image

**Describe the solution you'd like:** 
- [x] Fix the current starter pipeline to allow multiple variable groups
- [x] Modify the existing `services` azure-pipelines.yaml template to include the an ACR build job as the first build stage. The target ACR resource can be referenced from the pipeline variables
- [ ] Save the container tag/id for the second build stage - #647 

**Describe alternatives you've considered:**

**Additional context:**

**Does this require updates to documentation?:**

### PR DESCRIPTION
**Pull request by evanlouie**
This PR focuses around ironing out the flow of running `spk project init` or
`spk project init -m`, creating pipelines via the generated
`azure-pipelines.yaml` files, and ensuring the pipeline can create appropriately
named/tagged docker images in ACR.

In order to allow for flexibility in how SRE's choose to deploy the pipelines,
this PR does not automate the actual deployment of the pipeline. Instead, `info`
level instructions have been added specifying how to run the
`spk service create-pipeline` to deploy the `azure-pipelines.yaml` generated; If
the the user has configured the spk config correctly, the outputted command is
copy/paste ready to deploy the generated pipelines yaml:

If ran on a mono-repo:
`spk service create-pipeline --project-dir packages my-service` will be logged.

If ran on a standard repo: `spk service create-pipeline my-service` will be
logged.

The `azure-pipelines.yaml` expects 4 environment variables to be present in
order to login to `az` cli and build and push an image to ACR:

- `ACR_NAME` - name of the target ACR
- `SP_APP_ID` - service principal ID
- `SP_PASS` - service principal password
- `SP_TENANT` - service principal tenant

These requirements are also `info` logged to the end user upon
`azure-pipelines.yaml` generation.

Previously `spk service create-pipeline` only worked for mono-repositories and
the code has been refactored such that the existence of the `--packages-dir`
options implies whether or not the repository is a mono-repository or a standard
one.

closes https://github.com/microsoft/bedrock/issues/646

### CODE REVIEW

File: src/commands/service/pipeline.ts [hunk 1]
```diff
@@ -38,7 +37,7 @@ export const createPipelineCommandDecorator = (
     .option("-d, --devops-project <devops-project>", "Azure DevOps Project")
     .option(
       "-l, --packages-dir <packages-dir>",
-      "The monorepository directory containing this service definition. ie. '--packages-dir packages' if my-service is located under ./packages/my-service."
+      "The mono-repository directory containing this service definition. ie. '--packages-dir packages' if my-service is located under ./packages/my-service. Omitting this option implicitly means this is a standard repository."
```
Comment (evanlouie) [inline src/commands/service/pipeline.ts]: @bnookala note

File: src/commands/service/pipeline.ts [hunk 2]
```diff
@@ -166,16 +162,23 @@ export const installPipeline = async (
...
     /* tslint:disable-next-line object-literal-shorthand */
-    pipelineName: pipelineName,
-    repositoryName: repoName,
-    repositoryUrl: repoUrl,
+    pipelineName,
+    repositoryName,
+    repositoryUrl,
     yamlFileBranch: "master",
-    yamlFilePath: path.join(packagesDir, serviceName, "azure-pipelines.yaml") // This may not work if we're using a non-mono repository and azure-pipelines.yaml is in the root directory.
+    yamlFilePath: packagesDir // if a packages dir is supplied, its a mono-repo
```
Comment (evanlouie) [inline src/commands/service/pipeline.ts:169]: @bnookala note this new logic. Previously it only worked on mono-repos. This logic allows for standard as well

File: src/lib/fileutils.ts [hunk 3]
```diff
@@ -106,23 +99,56 @@ export const starterAzurePipelines = async (opts: {
...
       },
       ...cleanedPaths.map(projectPath => {
+        const projectPathParts = projectPath
+          .split(path.sep)
+          .filter(p => p !== "");
+        // If a the projectPath contains more than 1 segment (a service in a
+        // mono-repo), use the last as the project name as it will the folder
+        // containing the Dockerfile. Otherwise, its a standard service and does
+        // not need a a project name
+        const projectName =
```
Comment (evanlouie) [inline src/lib/fileutils.ts:134]: Important bit: in a mono-repo, we append `-<service-folder-name>` to the image name so services can be differentiated in the ACR. In the case of a standard repo, just the `Build.DefinitionName` is used

File: src/lib/fileutils.ts [hunk 4]
```diff
@@ -106,23 +99,56 @@ export const starterAzurePipelines = async (opts: {
...
+    const serviceName =
+      relPathParts.length > 1
+        ? relPathParts.slice(-2)[1]
+        : process
+            .cwd()
+            .split(path.sep)
+            .slice(-1)[0];
+    const spkServiceCreatePipelineCmd =
+      "spk service create-pipeline " + packagesOption + serviceName;
+    logger.info(
```
Comment (evanlouie) [inline src/lib/fileutils.ts:172]: Instructions are now outputted from `spk project init` on how to deploy the newly generated azure-pipelines.yaml via `spk service create-pipeline`

File: src/lib/pipelines/pipelines.ts [hunk 5]
```diff
@@ -212,10 +212,23 @@ export const createPipelineForDefinition = async (
...
   try {
-    return await buildApi.createDefinition(definition, azdoProject);
+    logger.debug(
+      `Creating BuildDefinition based on ${JSON.stringify(definition)}`
+    );
+    const createdDefn = await buildApi.createDefinition(
+      definition,
+      azdoProject
+    );
+    // type definition for createDefinition is wrong. It will resolve a `null` if an error occurs in azdo
```
Comment (evanlouie) [inline src/lib/pipelines/pipelines.ts:222]: @bnookala

**Review by mtarng — CHANGES_REQUESTED** [review]
Can you also update the appropriate docs under https://github.com/CatalystCode/spk/tree/master/docs ?

File: src/commands/service/pipeline.ts (hunk 2 above)
Comment (bnookala) [reply src/commands/service/pipeline.ts:169]: 👍

File: src/commands/service/pipeline.ts [hunk 6]
```diff
@@ -184,16 +187,18 @@ export const installPipeline = async (
     logger.error(err);
     return exitFn(1);
   }
+  if (typeof builtDefinition.id === "undefined") {
```
Comment (bnookala) [inline src/commands/service/pipeline.ts:190]: why would this be a special case error?

File: src/lib/pipelines/pipelines.ts (hunk 5 above)
Comment (bnookala) [reply src/lib/pipelines/pipelines.ts:222]: ooooh

**Review by bnookala — APPROVED** [review]

**Review by mtarng — APPROVED** [review]

**Pull request merged**